
## **Unique ID Generation for IoT Devices using Huffman Coding Technique**
#### _Suvam Basak_
#### _MTech IT_
#### _REG: 20MCMB08_

*********************************

List of MAC address of devices and the probability of sending messages.

In [1]:
# Python dictionary for MAC and the probability.

device_details = {
    '75:56:92:75:c7:7a':0.11,
    '92:6b:7e:82:3b:46':0.22,
    '0f:fe:0f:23:32:6a':0.16,
    'da:be:77:8e:5e:c3':0.12,
    '82:cd:f7:d7:25:27':0.15,
    'c6:ed:fb:56:22:92':0.10,
    '4c:5d:00:ef:a2:7f':0.14
}

print (device_details)


{'75:56:92:75:c7:7a': 0.11, '92:6b:7e:82:3b:46': 0.22, '0f:fe:0f:23:32:6a': 0.16, 'da:be:77:8e:5e:c3': 0.12, '82:cd:f7:d7:25:27': 0.15, 'c6:ed:fb:56:22:92': 0.1, '4c:5d:00:ef:a2:7f': 0.14}


Class of the node for each IoT devices in the network.

In [2]:
class HeapNode:
    '''
    This is the class of Devices. Takes the MAC address of devices and the probability.
    For leaf node left and right will be None.
    For internal node left and right will be holding the child node.
    '''
    def __init__(self, mac, probability):
        self.mac = mac
        self.probability = probability
        self.left = None
        self.right = None

    # Operator Overloading

    # lessthan will compare the probability of two object.
    def __lt__(self,other):
        return self.probability < other.probability
    
    # equal will check equality of probability of two object.
    def __eq__(self,other):
        if other == None:
            return False
        if not isinstance(other, HeapNode):
            return False
        return self.probability == other.probability

    # Show the details.
    def show_details(self):
        print('MAC : ',self.mac)
        print('PROB : ',self.probability)
        print('LEFT NODE : ',self.left)
        print('RIGHT NODE',self.right)
        

We need a Min Heap data structure for Huffman coding.

In [3]:
import heapq

# List for storing heap of HeapNode object.
device_heap = []


Creating the HeapNode for each device and adding it into the device heap.

In [4]:
for mac in device_details.keys():
    # Create HeapNode of each devices.
    new_node = HeapNode(mac,device_details[mac])
    # Adding into heap.
    heapq.heappush(device_heap,new_node)

print (device_heap)

[<__main__.HeapNode object at 0x7f20fc11fa90>, <__main__.HeapNode object at 0x7f20fc11f910>, <__main__.HeapNode object at 0x7f20fc11f7c0>, <__main__.HeapNode object at 0x7f20fc11f9d0>, <__main__.HeapNode object at 0x7f20fc11fa30>, <__main__.HeapNode object at 0x7f20fc11f940>, <__main__.HeapNode object at 0x7f20fc11fac0>]


Building the tree structure by:

1. Extracting two nodes (node1, node2) having minimal probability.
2. Create new node (new_node) with a probability = (node1 probability + node2 probability).
3. Setting node1 and node2 left and right child of new_node.
4. Adding the new_node into the heap.

In [5]:
while len(device_heap)>1:
    # Heapify the device_heap list.
    heapq.heapify(device_heap)
    # Extracting two minimal node.
    node1 = heapq.heappop(device_heap)
    node2 = heapq.heappop(device_heap)
    # Creating new node and setting left and right child.
    new_node = HeapNode(None,(node1.probability+node2.probability))
    new_node.left = node1
    new_node.right = node2
    # Adding new node into heap.
    heapq.heappush(device_heap,new_node)

# Showing details of root node.
head = device_heap[0]
head.show_details()

MAC :  None
PROB :  1.0
LEFT NODE :  <__main__.HeapNode object at 0x7f20fc11f400>
RIGHT NODE <__main__.HeapNode object at 0x7f20fc11f700>


Recursive function for generating the device IDs. The function takes root node and the current_id (by default empty string). For each non leaf node it recursively calls it self with left and right child. For left child it appends '0' to the current_id and '1' for right child.

In [6]:
# Python dictionary for storing the device IDs.
device_ids = {}

# Function to Generate the IDs.
def generate_id(root, current_id=''):
    # If the node is the leaf node.
    if (root.left == None) and (root.right == None):
        device_ids[root.mac]=current_id
        print(root.mac,current_id)
        return
    else:
        # recursive call for internal nodes.
        generate_id(root.left,current_id+'0')
        generate_id(root.right,current_id+'1')

# Calling the function with root node.
generate_id(head)

c6:ed:fb:56:22:92 000
75:56:92:75:c7:7a 001
92:6b:7e:82:3b:46 01
da:be:77:8e:5e:c3 100
4c:5d:00:ef:a2:7f 101
82:cd:f7:d7:25:27 110
0f:fe:0f:23:32:6a 111


All the generated IDs are stored in the dictionary device_ids.

In [7]:
print (device_ids)

{'c6:ed:fb:56:22:92': '000', '75:56:92:75:c7:7a': '001', '92:6b:7e:82:3b:46': '01', 'da:be:77:8e:5e:c3': '100', '4c:5d:00:ef:a2:7f': '101', '82:cd:f7:d7:25:27': '110', '0f:fe:0f:23:32:6a': '111'}


### Results

|     **MAC**     |**probability**|**ID**|
|-----------------|:-------------:|:----:|
|75:56:92:75:c7:7a|           0.11|   001|
|92:6b:7e:82:3b:46|           0.22|    01|
|0f:fe:0f:23:32:6a|           0.16|   111|
|da:be:77:8e:5e:c3|           0.12|   100|
|82:cd:f7:d7:25:27|           0.15|   110|
|c6:ed:fb:56:22:92|           0.10|   000|
|4c:5d:00:ef:a2:7f|           0.14|   101|